### ISSUE TITLE
<AmplifyTotpSetup> component doesn't render QR code

### ISSUE THREAD

**Issue by ronnyroeller** [aws-amplify/amplify-js#8575]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

Authentication

### Amplify Categories

auth

### Environment information

<details>

```
  System:
    OS: macOS 11.4
    CPU: (4) x64 Intel(R) Core(TM) i7-4558U CPU @ 2.80GHz
    Memory: 1.76 GB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.17.0 - ~/n/bin/node
    Yarn: 1.22.5 - ~/.yarn/bin/yarn
    npm: 6.14.11 - /usr/local/bin/npm
  Browsers:
    Brave Browser: 91.1.26.74
    Chrome: 91.0.4472.114
    Chrome Canary: 93.0.4571.0
    Edge: 91.0.864.37
    Firefox: 89.0
    Safari: 14.1.1
  npmPackages:
    @aws-amplify/ui-react: ^1.2.4 => 1.2.4 
    @testing-library/jest-dom: ^5.11.4 => 5.14.1 
    @testing-library/react: ^11.1.0 => 11.2.7 
    @testing-library/user-event: ^12.1.10 => 12.8.3 
    @types/jest: ^26.0.15 => 26.0.23 
    @types/node: ^12.0.0 => 12.20.15 
    @types/react: ^17.0.0 => 17.0.13 
    @types/react-dom: ^17.0.0 => 17.0.8 
    aws-amplify: ^4.1.2 => 4.1.2 
    react: ^17.0.2 => 17.0.2 
    react-dom: ^17.0.2 => 17.0.2 
    react-scripts: 4.0.3 => 4.0.3 
    typescript: ^4.1.2 => 4.3.5 
    web-vitals: ^1.0.1 => 1.1.2 
  npmGlobalPackages:
    @aws-amplify/cli: 4.37.0
    @collaborne/generator-carrot-service: 1.3.5
    @collaborne/generator-nodejs-microservice: 1.3.13
    @collaborne/generator-polaris-service: 2.3.2
    @collaborne/polaris-tenant-backup-service: 1.9.18
    @collaborne/polaris-token-service-tools: 1.2.0-0
    amplify: 0.0.11
    bower: 1.8.4
    browser-sync: 2.26.7
    cordova: 8.1.2
    cwtail: 1.3.0
    deploy: 1.0.3
    eslint: 4.19.1
    firebase-tools: 7.11.0
    gulp: 4.0.2
    ios-deploy: 1.9.4
    karma: 5.0.2
    miro-templating: 1.1.0
    mocha: 8.0.1
    n: 6.5.1
    near-cli: 2.0.1
    npm-check-updates: 2.14.2
    npm: 6.14.11
    parcel: 2.0.0-alpha.3.2
    plugman: 2.0.0
    pnpm: 5.5.2
    polymer-cli: 1.7.7
    postcss-cli: 7.1.1
    rollup: 1.17.0
    serve: 11.3.2
    ts-node: 8.10.2
    tslint: 5.11.0
    typescript: 4.1.2
    wasm-opt: 1.2.0
    yarn: 1.7.0
    yo: 3.1.1
```

</details>

### Describe the bug

The `<AmplifyTotpSetup>` component doesn't show the QR code.
![Screen+Shot+2021-07-03+at+10 04 26+PM (1)](https://user-images.githubusercontent.com/9339055/125210070-a28fad80-e29d-11eb-9295-3d80d5910a53.png)

### Expected behavior

The QR is shown as part of the `<AmplifyTotpSetup>` component.

### Reproduction steps

We created a minimal CRA app that allows to login and setup TOTP:
1. Create Cognito User Pool with MFA as optional and TOTP enabled
1. Install minimal reproduction CRA repo: https://github.com/ronnyroeller/amplify-totp
2. Adjust userPoolId/userPoolWebClientId for Auth.configure in App.tsx
3. Start CRA: `yarn start`
4. Create Cognito account
5. Try to set TOTP

Yet, the component doesn't show the QR code (see screenshot attached). When the user sends a random security token via the app, the Cognito responds with:
```json
{"__type":"SoftwareTokenMFANotFoundException","message":"Software Token MFA does not exist for the user."}
```

The documentation indicates that this exception means that TOTP-based MFA is not enabled for the user pool - yet, it's enabled as "optional" in the User Pool.

### Code Snippet

Repo with minimal reproduction case: https://github.com/ronnyroeller/amplify-totp

### Log output

<details>

```
There are no error/log messages.
```

</details>

### Additional information and screenshots

We also tried to create the SecretCode via the aws-cli, which does work:
```sh
$ aws cognito-idp associate-software-token --access-token XXX
{
    "SecretCode": "XXX"
}
```

**Comment by wlee221** [comment]
Hello @ronnyroeller,

Thanks for the repro! We were able to reproduce this locally. If I understand your use case correctly, you want to 

1) set the TOTP MFA as optional
2) on sign-in complete, trigger totp setup flow using `AmpliftyTotpSetup`

right?

**Comment by wlee221** [comment]
I took a look into the codebase. We don't have good support for optional TOTP (required TOTP works as expected because this triggers totp setup workflow by default). If you are planning on not requiring MFA and keep it optional, we would like to learn about your use case so we have better idea of how to resolve this. 

How can each user opt to use MFA? Will they be provided a prompt asking whether they want the TOTP enabled?

**Comment by ronnyroeller** [comment]
Indeed, we're using optional TOTP.

The user journey is as following:
1. User signs up to our app with email/password only
2. User can add MFA from the profile page for additional security (see screenshot below):
![image](https://user-images.githubusercontent.com/9339055/125603775-ab508d83-5f61-4bdd-a958-55e54c27f83e.png)

Today, we only support TOTP as a second factor but we might add more as Cognito starts supporting them (e.g. Yubikey, backup codes, etc).

This flows is comparable to e.g. Google where users can add further factors to secure their accounts (https://myaccount.google.com/security).

Note: The QR code used to be rendered correctly by the <AmplifyTotpSetup> component. We aren't sure when it stopped working.

**Comment by wlee221** [comment]
Ah, thanks for the clarification! I suspect https://github.com/aws-amplify/amplify-js/pull/7522 caused the misbehavior. 

This PR was made to accommodate a quite frustrating bug with web components... a slotted component (in this case `<AmplifyTotpSetup />`) always calls its `onMount` lifecycle method even before it's rendered. This is usually not a problem, but `AmplifyTotpSetup` in particular suffers from this bug because it runs TOTP setup and QR code generation, and obviously we don't want those to run before the element is visible. So #7522 alleviated this by only running setup when we're in the corresponding `AuthState`, ie. Cognito prompts us to do the TOTP challenge.

The problem with this PR is that now `<AmplifyTotpSetup />` no longer works standalone as it depends on the correct AuthState. (And to be transparent, `AmplifyTotpSetup` was not expected to be used directly by users).  Past few days I've been working on finding a solution that satisfies both standalone and internal use case, but I haven't had any luck due to how finicky the web component behaves with slots. 

Due to these nuances with web components, we're now rewriting our UI Components to not use web components. I'll make sure we cover this use case on our rewrite.

**Comment by wlee221** [comment]
For the actual issue... 😆 , I'm working on https://github.com/aws-amplify/amplify-js/pull/8592 that'll serve as a workaround while we work on the rewrite. This PR will ease the condition on `AmplifyTotpSetup` so that it can run when used externally too.

Only caveat with that PR is that you want the `<AmplifyTotpSetup />` outside the <AmplifyAuthenticator /> so that it's free of the bug above. For example, I was able to get it working with https://gist.github.com/wlee221/3361924b364974af42ce17d2fa53dc27.

I'll work on the PR so you can try this out soon.

**Comment by ronnyroeller** [comment]
This is fantastic, William. Thanks a lot!

**Comment by wlee221** [comment]
Hi @ronnyroeller, I just merged in #8592 that should resolve this PR. This is now on `unstable` npm tag and will be published in our next release cycle. I'll let you know again once it's released!

Please see the Usage Note on #8592. Essentially, adding `standalone` prop to `AmplifyTotpSetup` would enable the component to render the QR code, wherever it's placed at. Please try this out (on `unstable` or after it's released) and lets us know if this works!

**Comment by ronnyroeller** [comment]
That's excellent news, William. Thanks for making this happen!

**Comment by wlee221** [comment]
This now has been released to `@aws-amplify/ui-react@1.2.7` + `aws-amplify@4.2.1`. Let us know if the issue persists 👍

**Comment by ronnyroeller** [comment]
Thanks a lot, William. I just tried it out and can confirm that the issue is resolved.
